### PR TITLE
Fix gcc recipe for RHEL7.

### LINKED ISSUE #13391: gcc builds fail for C++ projects
It looks like #13202 introduced a libiconv dependency for gcc. This makes C++ builds fail for me with error messages such as these (in this case, cmake bootstrapping):
```
../opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/gcc-9.2.0-7kzpeggfu32layxcv6rzunmef5ah7uwv/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../lib64/libstdc++.so: undefined reference to `libiconv'
```
It looks like the binaries are correctly linked against libiconv but libstdc++ is not.

cc @adamjstewart

### LINKED ISSUE #13452: Installation issue: gperf@3.0.4 %gcc@9.2.0
### Steps to reproduce the issue

* git hash: 5582de837ef126ad736d44565a8d4231225b29da

```console
rm -rf ~/.spack
cd /scratch
rm -rf spack
git clone git@github.com:spack/spack && cd spack
export SPACK_ROOT=`pwd`
export PATH=`pwd`/bin:$PATH
spack compiler add
spack compiler list

==> Available compilers
-- gcc rhel7-x86_64 ---------------------------------------------
gcc@4.8.5

spack install gcc@9.2.0
source /scratch/spack/share/spack/setup-env.sh
spack load gcc@9.2.0
spack compiler add
spack compiler list

==> Available compilers
-- gcc rhel7-x86_64 ---------------------------------------------
gcc@9.2.0  gcc@4.8.5

spack install gperf % gcc@9.2.0

==> Installing gperf
==> Searching for binary cache of gperf
==> Warning: No Spack mirrors are currently configured
==> No binary for gperf found: installing from source
==> Fetching https://ftpmirror.gnu.org/gperf/gperf-3.0.4.tar.gz
######################################################################## 100.0%
==> Staging archive: /tmp/kellyt/spack-stage/spack-stage-gperf-3.0.4-buvuipk2igiid63wnhg2ckrdoayrpz7c/gperf-3.0.4.tar.gz
==> Created stage in /tmp/kellyt/spack-stage/spack-stage-gperf-3.0.4-buvuipk2igiid63wnhg2ckrdoayrpz7c
==> No patches needed for gperf
==> Building gperf [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

6 errors found in build log:
     95     /scratch/kellyt/spack.20191025/lib/spack/env/gcc/g++ -g -O2 -I. -I.
            /../lib -c ./bool-array.cc
     96     /scratch/kellyt/spack.20191025/lib/spack/env/gcc/g++ -g -O2 -I. -I.
            /../lib -c ./hash-table.cc
     97     /scratch/kellyt/spack.20191025/lib/spack/env/gcc/g++ -g -O2 -I. -I.
            /../lib -c ./search.cc
     98     /scratch/kellyt/spack.20191025/lib/spack/env/gcc/g++ -g -O2 -I. -I.
            /../lib -c ./output.cc
     99     /scratch/kellyt/spack.20191025/lib/spack/env/gcc/g++ -g -O2 -I. -I.
            /../lib -c ./main.cc
     100    /scratch/kellyt/spack.20191025/lib/spack/env/gcc/g++ -g -O2  -o gpe
            rf version.o positions.o options.o keyword.o keyword-list.o input.o
             bool-array.o hash-table.o search.o output.o main.o ../lib/libgp.a 
            -lm
  >> 101    /scratch/kellyt/spack.20191025/opt/spack/linux-rhel7-haswell/gcc-4.
            8.5/gcc-9.2.0-7qmgg3c3xt2bpg4typfxg7imt5zgpi3i/lib/gcc/x86_64-pc-li
            nux-gnu/9.2.0/../../../../lib64/libstdc++.so: undefined reference t
            o `libiconv'
  >> 102    /scratch/kellyt/spack.20191025/opt/spack/linux-rhel7-haswell/gcc-4.
            8.5/gcc-9.2.0-7qmgg3c3xt2bpg4typfxg7imt5zgpi3i/lib/gcc/x86_64-pc-li
            nux-gnu/9.2.0/../../../../lib64/libstdc++.so: undefined reference t
            o `libiconv_close'
  >> 103    /scratch/kellyt/spack.20191025/opt/spack/linux-rhel7-haswell/gcc-4.
            8.5/gcc-9.2.0-7qmgg3c3xt2bpg4typfxg7imt5zgpi3i/lib/gcc/x86_64-pc-li
            nux-gnu/9.2.0/../../../../lib64/libstdc++.so: undefined reference t
            o `libiconv_open'
  >> 104    collect2: error: ld returned 1 exit status
```

### Platform and user environment

Please report your OS here:
```commandline
$ uname -a 
Linux x.x.x 3.10.0-1062.1.2.el7.x86_64 #1 SMP Mon Sep 16 14:19:51 EDT 2019 x86_64 x86_64 x86_64 GNU/Linux

$ lsb_release -d
Description:	Red Hat Enterprise Linux Workstation release 7.7 (Maipo)

% spack arch
linux-rhel7-broadwell
``` 

config files
- [compilers.yaml.txt](https://github.com/spack/spack/files/3774427/compilers.yaml.txt)
- I don't have a packages.yaml file.

### Additional information

Sometimes the issue benefits from additional details. In these cases there are
a few things we can suggest doing. First of all, you can post the full output of:

```console
$ spack spec --install-status  gperf
Input spec
--------------------------------
 -   gperf

Concretized
--------------------------------
 -   gperf@3.0.4%gcc@9.2.0 arch=linux-rhel7-broadwell
```

to show people whether Spack installed a faulty software or if it was not able to
build it at all. 

```console
$ spack location -s gperf
/tmp/$USER/spack-stage/spack-stage-gperf-3.0.4-buvuipk2igiid63wnhg2ckrdoayrpz7c
```
Rebuilding the package with the following options:
```console
$ spack -d install -j 1 gperf
# produced the same error as above.
```

### PR DESCRIPTION
**Pull request by KineticTheory**
+ It appears that macOS related changes to the gcc recipe (PR #13202) broke gcc on RHEL7.  This bug manifests as `libstdc++.so: undefined reference to libiconv` when gcc is used.
+ Fixes #13452 by moving `--with-libiconv-prefix={0}'.format(spec['libiconv'].prefix)` into the darwin OS section of the configuration.
+ Also fixes #13391
+ Needs to be tested on macOS and other Linux flavors.

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -100,7 +100,7 @@ class Gcc(AutotoolsPackage):
     depends_on('isl@0.15:0.18', when='@6:8.9')
     depends_on('isl@0.15:0.20', when='@9:')
     depends_on('zlib', when='@6:')
-    depends_on('libiconv')
+    depends_on('libiconv', type='build')
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py]: To make sure that the build system does not pick it up on non-Darwin:
```suggestion
    depends_on('libiconv', type='build', when='platform=darwin')
```

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/gcc/package.py]: Shouldn’t be build type because it links to the library

Comment (maiterth) [reply var/spack/repos/builtin/packages/gcc/package.py]: I had the same issue on my ubuntu19.04 machine trying to build gcc@9.2.0%gcc@9.2.0. 
The patch fixes the issues and I can compile.

Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/gcc/package.py]: I also just recompiled our software stack and this PR (plus my suggested change) fixes the problems I was having with GCC.

Comment (KineticTheory) [reply var/spack/repos/builtin/packages/gcc/package.py]: Requested changes to `depends_on` line are in commit 08a06a7.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 2]
```diff
@@ -328,7 +327,8 @@ def configure_args(self):
         if sys.platform == 'darwin':
             options.extend([
                 '--with-native-system-header-dir=/usr/include',
-                '--with-sysroot={0}'.format(macos_sdk_path())
+                '--with-sysroot={0}'.format(macos_sdk_path()),
+                '--with-libiconv-prefix={0}'.format(spec['libiconv'].prefix),
```
Comment (pat-s) [inline var/spack/repos/builtin/packages/gcc/package.py]: Is the trailing comma here on purpose?

Comment (KineticTheory) [reply var/spack/repos/builtin/packages/gcc/package.py]: Dang it!  Cut & paste error.  Thanks for spotting that issue. Update is incoming...

**Comment by alalazo** [comment]
@KineticTheory Can you please turn the deprecated API call: 
```python
setup_environment(self, build_env, run_env)
```
into:
```python
setup_run_environment(self, run_env)
```
? This is what the failing test is asking for and is due to #11115 - not anything directly related to this PR. I proposed in that PR  to have a test that would fail if any package modified in a PR contains deprecated methods. The intent is to let the community know that a new API is available. Since it's a first time for something like that any feedback or thought is welcome.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 3]
```diff
@@ -424,7 +424,7 @@ def write_rpath_specs(self):
                               self.prefix.lib, self.prefix.lib64))
         set_install_permissions(specs_file)
 
-    def setup_environment(self, spack_env, run_env):
+    def setup_run_environment(self, spack_env, run_env):
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/gcc/package.py]: The new call has been split into two calls. Therefore the signature takes only one environment:
```python
def setup_run_environment(self, env):
```

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 4]
```diff
@@ -424,7 +424,7 @@ def write_rpath_specs(self):
                               self.prefix.lib, self.prefix.lib64))
         set_install_permissions(specs_file)
 
-    def setup_environment(self, spack_env, run_env):
+    def setup_run_environment(self, run_env):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gcc/package.py]: I would rename `run_env` to `env` in this method to be consistent with other uses.

File: var/spack/repos/builtin/packages/gcc/package.py (hunk 3 above)
Comment (KineticTheory) [reply var/spack/repos/builtin/packages/gcc/package.py]: Sorry about that.  Should be fixed in e2f42af.  Let's see what Travis has to say...

**Pull request merged**